File: fortran/src/collisions.f90
!---------------------------------------------------------------------!
! GCAF= Guiding Center Analytic Field code by R. Saavedra             !
! Email of the author: saavestro@gmail.com                            !
! For more information please visit https://github.com/Saavestro/gcaf !
!---------------------------------------------------------------------!
!---------------------------------------------------------!
subroutine scatr(k)
!-----------------------------------------------------!
! Boozer and Kuo-Petravic Phys. Fluids 24, 851 (1981) !
!-----------------------------------------------------!
!=========================================================!
	use shared,only: &
		pi,&     !pi number
		keverg,& !keV to erg
		rq,&     !elec charge
		mp,&     !proton mass
		ndens,&  !dens profile flat/para
		ntemp,&  !temp profile flat/gaus
		ntmpi,&	 !input ions temp in keV
		ntmpe,&	 !input elec temp in keV
		deno,&   !plasma dens in cm^-3
		ptch,&   !pitch angle
		en,&     !energy
		pol,&    !poloidal flux
		rho,&    !parallel gyroradius
		polw,&   !on wall poloidal flux
		zi,&	 !ions charge in protons charge
		zt,&     !test charge in protons charge
		mt,&     !test particle mass in protons mass
		mi,&     !ions mass in protons mass
		enaxo,&  !on axis particle en 
		leno,&   !characteristic length
		omego,&  !characteristic time freq
		dt,&     !time step
		bf,&     !b field 
		pt,&     !electrostatic potential
		mu       !b moment
	implicit none
	integer(4):: k
	real(8):: dni       !ions density in cm-3
	real(8):: ti        !ions temperature in code units
	real(8):: te        !electrons temperature code units
	real(8):: vti       !ions thermal velocity
	real(8):: dum
	real(8):: fi
	real(8):: gi
	real(8):: efaci
	real(8):: eface
	real(8):: psig      !pitch operator random sign
	real(8):: esig      !en operator random sign
	real(8):: xd		!pol/polw for dens_prof
	real(8):: xt		!pol/polw for temp_prof
	real(8):: vprt      !test particle thermal velocity
	real(8):: dens_prof !dens profile function
	real(8):: temp_prof !temp profile function
	real(8):: ranx      !random function
	real(8):: vte       !electron thermal velocity
	real(8):: fe
	real(8):: ge
	real(8):: dumi      !vprt/vti
	real(8):: dume      !vprt/vte
	real(8):: clogi     !ions Coulomb log
	real(8):: cloge     !elec Coulomb log
	real(8):: tikev     !ions temp in KeV
	real(8):: tekev     !elec temp in KeV
	real(8):: dne       !elec dens in cm-3
	real(8):: eekev     !test kinetic en in keV
	real(8):: ekin      !test kinetic en
	real(8):: ekmin     !test en minimum
	real(8):: colbi     !ions braginskii freq
	real(8):: colbe     !elec braginskii freq
	real(8):: coldi     !ions deflection freq
	real(8):: colde     !elec deflection freq
	real(8):: colei     !ions en freq
	real(8):: colee     !elec en freq
	real(8):: colds     !deflection freq sum
	real(8):: coles     !en freq sum
	real(8):: velo      !characteristic velocity
	real(8):: clogtest 
	real(8):: bmintest
	real(8):: bmaxtest
	real(8):: mutest
	real(8):: usqtest
!=========================================================!

	!SET PARAMETERS
	xd= pol(k)/polw
	if(ndens.eq.0) xd= 0.d0
	xt= pol(k)/polw
	if(ntemp.eq.0) xt= 0.d0
	ekin= en(k)
	eekev= enaxo*en(k)/keverg 
	dni= dens_prof(xd,deno)
	dne= dni*zi 
	tikev= temp_prof(xt,ntmpi/ntmpi)
	tekev= temp_prof(xt,ntmpe/ntmpi)
	ti= keverg*tikev/enaxo 
	te= keverg*tekev/enaxo 
	ekmin= 1.d-3*ti
	vprt= sqrt(2*ekin/mt)
	vti= sqrt(ti/mi)
	vte= sqrt(1836*te) 
	velo= leno*omego
	
	!IONS
	dumi= vprt/vti
	fi= ((1.d0- .5d0/dumi/dumi)*erf(dumi) &
		+exp(-dumi*dumi)/sqrt(pi)/dumi)/dumi/dumi/dumi
	gi = .5d0*(erf(dumi) -2.d0*dumi*exp(-dumi**2)/sqrt(pi))/dumi**3 !en
	efaci= ekin -exp(-dumi**2)*ti/gi/sqrt(pi) !en
	clogi= 23.d0 -log(zt*zt*zt*sqrt(2*dni)/(1.d+3*tikev)**1.5d0)
	!!
	mutest= mp*mi*mt/(mi +mt)
	usqtest= leno*leno*omego*omego*vprt*vprt
	bmintest= zt*zi*rq*rq/mutest/usqtest
	bmaxtest= 1/sqrt(dni*zi*zi*rq*rq/tikev/keverg +dne*rq*rq/tekev/keverg)
	clogtest= log(bmaxtest/bmintest)
	!!
	colbi= 1.2717d+11*(1.d0*zi*zi*zi/mi/mi)*dni*clogi &
		/velo/velo/velo/vti/vti/vti
	coldi= abs(1.8799d0*(zt*zt*mi*mi/zi/zi*mt*mt)*colbi*fi)/omego
	colei= abs(3.7599d0*(zt**2*mi/zi**2/mt)*colbi*gi)/omego !en

	!ELECTRONS
	dume= vprt/vte
	fe= ((1.d0- .5d0/dume**2)*erf(dume) &
		+exp(-dume**2)/sqrt(pi)/dume)/dume**3
	ge= .5d0*(erf(dume) -2.d0*dume*exp(-dume**2)/sqrt(pi))/dume**3 !en
	eface= ekin -exp(-dume**2)*te/ge/sqrt(pi) !en
	if(1.d+3*tekev.le. 5.46d-4*zt*1.d+3*tikev/mt) &
		cloge= 23.d0 -log(zt*sqrt(dne)/(1.d+3*tekev)**1.5d0)
	if(1.d+3*tekev.gt. 5.46d-4*zt*1.d+3*tikev/mt .and. &
		1.d+3*tekev.le. 10.d0*zt**2) &
		cloge= 30.d0 -log(zt*sqrt(dne)/(1.d+3*tekev)**1.5d0/mt)
	if(1.d+3*tekev.gt. 10.d0*zt**2) &
		cloge= 24.d0 -log(sqrt(dne)/1.d+3/tekev)
	colbe= 6.0634d+17*zi**2*dne*cloge/(leno*omego)**3/vte**3
	colde= abs(1.3293d0*(zt**2/zi/mi**2)*colbe*fe/1836**2)/omego
	colee= abs(2.6586d0*(zt**2/zi**2/mt)*colbe*ge/1836)/omego !en

	!FREQUENCIES
	colds= coldi +colde
	coles= ti*colei +te*colee !en
	
	!COLLISION STEP
	!pitch-angle step 
	dum= max(0.d0,1.d0- ptch(k)**2)
	psig= sign(1.d0,ranx()-.5d0)*sqrt(dum*colds*dt(k))
	ptch(k)= ptch(k)*(1.d0 -colds*dt(k))
	ptch(k)= ptch(k) +psig
	ptch(k)= min(1.d0,ptch(k))
	ptch(k)= max(-1.d0,ptch(k))
	!en step
	esig= sign(1.d0,ranx()-.5d0)*2.d0*sqrt(coles*ekin*dt(k))
	ekin= ekin -2.d0*colei*efaci*dt(k) -2.d0*colee*eface*dt(k)
	ekin= ekin +esig
	ekin= max(ekin,ekmin)
	!en(k)= ekin 
	
	!change rho, mu, due to scattering
	rho(k)= ptch(k)*sqrt(2*en(k))/bf(k)
	mu(k)= en(k)/bf(k)- .5d0*rho(k)*rho(k)*bf(k)
	return
end subroutine
!---------------------------------------------------------!


!---------------------------------------------------------!
function dens_prof(xd,den)
!=========================================================!
	implicit none
	real(8):: dens_prof !dens_prof function
	real(8):: den       !input density
	real(8):: xd        !variable
!=========================================================!
	dens_prof= den*(1.d0- xd*xd)
	return
end function
!---------------------------------------------------------!


!---------------------------------------------------------!
function temp_prof(xd,temp)
!=========================================================!
	use shared,only: &
		pi
	implicit none
	integer(8):: temp   !input temperature	
	real(8):: temp_prof !temp_prof function
	real(8):: xd     	!variable
	real(8):: kern    	!gaussian kernel
!=========================================================!
	kern= xd
	temp_prof= temp*exp(-kern)
	return
end function
!---------------------------------------------------------!

File: cpp/collisions.cpp
//----------------------------------------------------------------//
// gcafpp code by R. Saavedra. Email: saavestro@gmail.com         //
// For more information visit https://git.rsaavedra.xyz/?p=gcaf.git //
//----------------------------------------------------------------//
#include "shared.h"
const double signnum(const double&);


const double dens_prof(
	const double& xd){ // density parabolic profile
	return 1.0- xd*xd;
}

const double temp_prof(
	const double& xd){ // temperature gaussian profile
	return exp(-xd);
}

const double deflection_factor(
	const double& x){ // deflection frequency factor
	return (1.0 -0.5/x/x)*erf(x)/x/x/x +exp(-x*x)/sqrt(M_PI)/x/x/x/x;
}

const double energy_factor(
	const double& x){ // energy frequency factor
	return 0.5*erf(x)/x/x/x -exp(-x*x)/sqrt(M_PI)/x/x;
}

const std::array<double,3> white(
	const double& sqx){ // White energy factor
	const double a_0= 0.75225;
	const double a_1= -0.238;
	const double a_2= 0.311;
	const double a_3= -0.0956;
	const double a_4= 0.0156;
	const double x= sqx*sqx;
	const double coeffs=
		a_0 +a_1*x +a_2*x*x +a_3*x*x*x +a_4*x*x*x*x;
	const double pfun=
		sqx*sqx*sqx*coeffs;
	const double psi_fun_approx=
		1.0 -1.0/(1.0 +pfun);
	const double psi_fun_prime=
		2*sqx*exp(-x)/sqrt(M_PI);
	const double psi_fun_prime_prime=
		exp(-x)/sqx/sqrt(M_PI)
		-2*sqx*exp(-x)/sqrt(M_PI);

	return {psi_fun_approx, psi_fun_prime, psi_fun_prime_prime};
}

void scatter(
	const int& k,
	const double& ran_num_1,
	const double& ran_num_2){

	const double ap0= 0.75225;
	const double ap1= -0.238;
	const double ap2= 0.311;
	const double ap3= -0.0956;
	const double ap4= 0.0156;

	const double prot= // test mass
		(double)test.i_mass;
	const double zprt= // test charge
		(double)test.i_chrg;
	const double massb= // background mass
		(double)plasma.i_mass;
	const double chgb= // background charge
		(double)plasma.i_chrg;

	// parameters
	const double ekmin= // minimum test kinetic energy in code units
		1.0e-2*KEV_TO_ERG*code.d_en/code.d_enax;
	double ekin= // test kinetic energy in code units
		en[k] -pt[k];
	const double eekev= // test kinetic energy in keV
		ekin*code.d_enax/KEV_TO_ERG;
	double dnb; // background density in cm^-3
	switch (plasma.dens_prof) {
		case DEN_FLAT: {
				dnb= plasma.d_dens;
				break;
			}
		case DEN_PARA: {
				dnb= plasma.d_dens
					*dens_prof(pol[k]/code.d_polw);
				break;
			}
	}
	const double dne= // electrons density
		dnb*chgb;
	double tikev; // background temperature in keV
	switch (plasma.temp_prof) {
		case TEM_FLAT: {
				tikev= (double)plasma.i_temp/1000;
				break;
			}
		case TEM_GAUS: {
				tikev= (double)plasma.i_temp/1000
					*temp_prof(pol[k]/code.d_polw);
				break;
			}
	}
	const double ti=
		tikev*KEV_TO_ERG/code.d_enax;
	const double tekev=
		tikev;
	const double te=
		tekev*KEV_TO_ERG/code.d_enax;
	const double vprt= // test particle velocity
		sqrt(2*ekin);
	const double vtb= // background velocity
		sqrt(2*ti/massb);
	const double vte= // electrons velocity
		sqrt(2*te*1836);

	const double cnst= // Coulomb logarithm units
		2.41e+11*(zprt/prot)*(zprt/prot)
		/(code.d_leng*code.d_freq*vprt)
		/(code.d_leng*code.d_freq*vprt)
		/(code.d_leng*code.d_freq*vprt);

	{ // background
		const double dnu0= // background species s^-1
			cnst*dnb*chgb*chgb;
		const double dumb= // velocity ratio
			vprt/vtb;
		const double dd= // dd= x, dumb= sqrt(x)
			dumb*dumb;
		const double d3=
			dumb*dumb*dumb;
		const double dum=
			d3*(ap0 +ap1*dd +ap2*dd*dd +ap3*dd*dd*dd +ap4*dd*dd*dd*dd);
		const double psib= // approx to psi fnctn
			1.0 -1.0/(1.0 +dum);
		const double fb=
			(2.0 -1.0/dd)*psib +2.257*dumb*exp(-dd);
		const double gb=
			2*chgb*psib/massb -2.257*dumb*exp(-dd);
		const double gbp=
			2.257*exp(-dd)*((1.0 +chgb/massb)*dumb -0.5/dumb);

		// find coll freq and convert from sec-1 to code units
		const double bmx=
			7.4e-3*sqrt(tikev*1.0e+13/(zprt*zprt*dnb));
		const double bmin1=
			1.4e-10*zprt*chgb/(eekev +tikev);
		const double massab=
			prot*massb/(prot +massb);
		const double bmin2=
			6.2e-10/(sqrt(massab*1836*tikev));
		const double bdum=
			std::max(bmin1,bmin2);
		const double clogb= // ib Coulomb log
			log(bmx/bdum);
		const double colb= // scatter
			abs(clogb*dnu0*fb)/code.d_freq;
		const double colbs= // slow
			abs(clogb*dnu0*gb)/code.d_freq;

		// scatter and drag
		{ // pitch angle
			const double dum=
				std::max(0.0,1.0 -ptch[k]*ptch[k]);
			const double psig=
				signnum(ran_num_1 -0.5)*sqrt(dum*colb*dt[k]*0.5);
			ptch[k]=
				ptch[k]*(1.0 -colb*dt[k]*0.5) +psig;
		}
		{ // energy
			const double esig=
				signnum(ran_num_2 -0.5)*sqrt(2*ekin*ti*colbs*dt[k]);
			const double efac0=
				ekin*(1.0 -massb*gbp/(gb*prot));
			ekin=
				ekin -colbs*dt[k]*efac0 +esig;
			ekin=
				std::max(ekin,ekmin); // truncate slowing down;
		}
	}

	{ // electrons
		const double enu0= // electrons  sec-1
			cnst*dne;
		const double dume=
			vprt/vte;
		const double dd=
			dume*dume;
		const double d3=
			dume*dume*dume;
		const double dum=
			d3*(ap0 +ap1*dd +ap2*dd*dd +ap3*dd*dd*dd +ap4*dd*dd*dd*dd);
		const double psie= // approx to psi fnctn
			1.0 -1.0/(1.0 +dum);
		const double fe=
			(2.0 -1.0/dd)*psie +2.257*dume*exp(-dd);
		const double ge=
			2*prot*psie*1836 -2.257*dume*exp(-dd);
		const double gep=
			2.257*exp(-dd)*((1.0 +prot*1836)*dume -0.5/dume);

		// find coll freq and convert from sec-1 to code units
		const double bmx=
			7.4e-3*sqrt(tekev*1.0e+13/(zprt*zprt*dne));
		const double bmin1=
			1.4e-10*zprt/(eekev +tekev);
		const double massab=
			prot/(1836*prot +1.0);
		const double bmin2=
			6.2e-10/(sqrt(massab*1836*tekev));
		const double bdum=
			std::max(bmin1,bmin2);
		const double cloge= // ie Coulomb log
			log(bmx/bdum);
		const double cole= // scatter
			abs(cloge*enu0*fe)/code.d_freq;
		const double coles= // slow
			abs(cloge*enu0*ge)/code.d_freq;

		// scatter and drag
		{ // pitch angle
			const double dum=
				std::max(0.0,1.0 -ptch[k]*ptch[k]);
			const double psig=
				signnum(ran_num_1 -0.5)*sqrt(dum*cole*dt[k]*0.5);
			ptch[k]=
				ptch[k]*(1.0 -cole*dt[k]*0.5) +psig;
		}
		{ // energy
			const double esig=
				signnum(ran_num_2 -0.5)*sqrt(2*ekin*te*coles*dt[k]);
			const double efac0=
				ekin*(1.0 -gep/(ge*1836*prot));
			ekin=
				ekin -coles*dt[k]*efac0 +esig;
			ekin=
				std::max(ekin,ekmin); // truncate slowing down;
		}
	}

	ptch[k]= std::min(1.0,ptch[k]);
	ptch[k]= std::max(-1.0,ptch[k]);
	en[k]= ekin +pt[k];

	rho[k]= ptch[k]*sqrt(2*ekin)/bf[k];
	mu[k]= ekin/bf[k] -0.5*rho[k]*rho[k]*bf[k];
}


void scatter_old(
	const int& k,
	const double& ran_num_1,
	const double& ran_num_2){
	// test particle parameters
	double mass_tst; // test mass in grams
	double chrg_tst; // test charge in statC
	switch (test.sign) {
		case CHG_IONS:
		{
			mass_tst= MASS_PROT*test.i_mass;
			chrg_tst= CHRG_PROT*test.i_chrg;
			break;
		}
		case CHG_ELEC:
		{
			mass_tst= MASS_ELEC;
			chrg_tst= CHRG_PROT;
			break;
		}
	}
	const double vel_tst=  // test particle speed in cm*s^-1
		sqrt(2*code.d_enax*en[k]/mass_tst);
	double ekin= en[k] -pt[k]; // test particle kinetic energy

	const int rand_sign_1= // random sign
		signnum(ran_num_1 -0.5);
	const int rand_sign_2= // random sign
		signnum(ran_num_2 -0.5);
	double pch_slow_sum= 0.0; // pitch angle slowdown term sum
	double pch_scat_sum= 0.0; // pitch angle scatter term sum
	double eng_slow_sum= 0.0; // energy slowdown term sum
	double eng_scat_sum= 0.0; // energy scatter term sum
for(int i=0;i<total_species;i++){
	switch (species[i].sign){
		case CHG_IONS:
		{
			// ions parameters
			const double mass_ion= // mass in g
				MASS_PROT*species[i].i_mass;
			const double chrg_ion= // charge in statC
				CHRG_PROT*species[i].i_chrg;
			double temp_ion; // temperature in ergs
			switch (species[i].temp_prof) {
				case TEM_FLAT: {
						temp_ion= KEV_TO_ERG*species[i].i_temp;
						break;
					}
				case TEM_GAUS: {
						temp_ion= KEV_TO_ERG*species[i].i_temp
							*temp_prof(pol[k]/code.d_polw);
						break;
					}
			}
			double dens_ion; // density in cm^-3
			switch (species[i].dens_prof) {
				case DEN_FLAT: {
						dens_ion= species[i].d_dens;
						break;
					}
				case DEN_PARA: {
						dens_ion= species[i].d_dens
							*dens_prof(pol[k]/code.d_polw);
						break;
					}
			}
			const double vel_ion=  // speed in cm*s^-1
				sqrt(3*temp_ion/mass_ion);
			const double x_ion= // velocity ratio
				vel_tst/vel_ion;

			const double bmin_ion= // impact parameter in cm
				chrg_tst*chrg_ion
				*(mass_tst +mass_ion)/mass_tst/mass_ion
				/vel_ion/vel_ion;
			const double bmax_ion= // debye length in cm
				sqrt(temp_ion/dens_ion/chrg_ion/chrg_ion);
			const double clog_ion= // coulomb logarithm
				log(bmax_ion/bmin_ion);
			const double nu_bi=    // braginskii frequency
				1.3333*sqrt(2*2*2*M_PI)
				*chrg_tst*chrg_tst*chrg_ion*chrg_ion/mass_tst/mass_ion
				*dens_ion*clog_ion
				/vel_ion/vel_ion/vel_ion/code.d_freq;

			// deflection operator
			const double coef_di= // deflection coefficient
				1.5*sqrt(0.5*M_PI)
				*chrg_tst*chrg_tst*mass_ion*mass_ion
				/chrg_ion/chrg_ion/mass_tst/mass_tst;
			const double nu_di= // deflection frequency
					coef_di*nu_bi*deflection_factor(x_ion);
			pch_slow_sum+= // pitch angle slowdown term
				-nu_di*ptch[k]*dt[k];
			pch_scat_sum+= // pitch angle scatter term
				(1.0- ptch[k]*ptch[k])*nu_di*dt[k];

			// energy operator
			const double coef_ei= // energy coefficient
				3*sqrt(M_PI/2)
				*mass_ion*chrg_tst*chrg_tst
				/chrg_ion/chrg_ion/mass_tst;
			const double nu_ei= // energy frequency
				coef_ei*nu_bi*energy_factor(x_ion);
			// const double dum=
			// 	ekin -1.5*mass_ion*exp(-x_ion*x_ion)*temp_ion
			// 	/energy_factor(x_ion)/mass_tst/sqrt(M_PI)/code.d_enax;
			const std::array<double,3> psi_fun= white(x_ion);
			const double dum=
				mass_ion*(psi_fun[0] +psi_fun[1])/mass_tst
				-psi_fun[1] -psi_fun[2];
			eng_slow_sum+= // energy slowdown term
				-2*nu_ei*ekin*dum*dt[k];
			eng_scat_sum+= // energy scatter term
				nu_ei*temp_ion*ekin*dt[k]/code.d_enax;
			break;
		}
		case CHG_ELEC:
		{
			// electrons parameters
			const double mass_ele= // mass in g
				MASS_ELEC*species[i].i_mass;
			const double chrg_ele= // charge in statC
				CHRG_PROT*abs(species[i].i_chrg);
			double temp_ele; // temperature in ergs
			switch (species[i].temp_prof) {
				case TEM_FLAT: {
						temp_ele= KEV_TO_ERG*species[i].i_temp;
						break;
					}
				case TEM_GAUS: {
						temp_ele= KEV_TO_ERG*species[i].i_temp
							*temp_prof(pol[k]/code.d_polw);
						break;
					}
			}
			double dens_ele; // density in cm^-3
			switch (species[i].dens_prof) {
				case DEN_FLAT: {
						dens_ele= species[i].d_dens;
						break;
					}
				case DEN_PARA: {
						dens_ele= species[i].d_dens
							*dens_prof(pol[k]/code.d_polw);
						break;
					}
			}
			const double vel_ele=  // speed in cm*s^-1
				sqrt(3*temp_ele/mass_ele);
			const double x_ele= // velocity ratio
				vel_tst/vel_ele;

			const double bmin_ele= // electrons impact parameter in cm
				chrg_tst*chrg_ele
				*(mass_tst +mass_ele)/mass_tst/mass_ele
				/vel_ele/vel_ele;
			const double bmax_ele= // electrons debye length in cm
				sqrt(temp_ele/dens_ele/chrg_ele/chrg_ele);
			const double clog_ele= // electrons coulomb logarithm
				log(bmax_ele/bmin_ele);
			const double nu_be= // braginskii electrons frequency
				5.3333*sqrt(M_PI)
				*chrg_tst*chrg_tst*chrg_ele*chrg_ele/mass_ele/mass_ele
				*dens_ele*clog_ele
				/vel_ele/vel_ele/vel_ele/code.d_freq;

			// deflection operator
			const double coef_de=
				0.75*sqrt(M_PI)
				*chrg_tst*chrg_tst*mass_ele*mass_ele
				/chrg_ele/chrg_ele/mass_tst/mass_tst;
			const double nu_de=
				coef_de*nu_be*deflection_factor(x_ele);
			pch_slow_sum+= // pitch angle slowdown term
				-nu_de*ptch[k]*dt[k];
			pch_scat_sum+= // pitch angle scatter term
				(1.0- ptch[k]*ptch[k])*nu_de*dt[k];

			// energy operator
			const double coef_ee=
				1.5*sqrt(M_PI)
				*mass_ele*chrg_tst*chrg_tst/mass_tst/chrg_ele/chrg_ele;
			const double nu_ee=
				coef_ee*nu_be*energy_factor(x_ele);
			// const double dum=
			// 	ekin -1.5*mass_ele*exp(-x_ele*x_ele)*temp_ele
			// 	/energy_factor(x_ele)/mass_tst/sqrt(M_PI)/code.d_enax;
			const std::array<double,3> psi_fun= white(x_ele);
			const double dum=
				mass_ele*(psi_fun[0] +psi_fun[1])/mass_tst
				-psi_fun[1] -psi_fun[2];
			eng_slow_sum+= // energy slowdown term
				-2*nu_ee*ekin*dum*dt[k];
			eng_scat_sum+= // energy scatter term
				nu_ee*temp_ele*ekin/code.d_enax*dt[k];
			break;
		}
	}
}

	{ // pitch-angle collision step
		pch_scat_sum= // scatter term
			rand_sign_1*sqrt(pch_scat_sum);
		ptch[k]+= pch_slow_sum +pch_scat_sum;
		ptch[k]= std::min(1.0,ptch[k]);
		ptch[k]= std::max(-1.0,ptch[k]);
	}

	{ // energy collision step
		const double ekin_min= // minimum energy bound
			1.0e-2*1*KEV_TO_ERG/code.d_enax;
		eng_scat_sum= // scatter term
			rand_sign_2*2*sqrt(eng_scat_sum);
		ekin+= eng_slow_sum +eng_scat_sum;
		ekin= std::max(ekin,ekin_min); // truncate slowdown
		en[k]= ekin;
	}

	{ // update variables
		rho[k]= ptch[k]*sqrt(2*en[k])/bf[k];
		mu[k]= en[k]/bf[k] -0.5*rho[k]*rho[k]*bf[k];
	}
}
